### PR TITLE
Stamp People [Second Try]

### PR DESCRIPTION
**Pull request by kugamo**
## What Does This PR Do
This PR adds the ability to stamp people with stamps. They will get a colored mark on them in the location where the stamper stamps them. The marks can be washed off with all the regular washing methods. Examination of the stamped person will show the stamp's unique marks in text. This can only be done to humans/human-like species.

This PR was opened due to the fact that the original https://github.com/ParadiseSS13/Paradise/pull/14628 had problems, I have taken actions in an attempt to resolve them. The code is unchanged apart from the addition of a count check, preventing no more than 20 marks from appearing on a person.

## Why It's Good For The Game
General fun. Collect them all and become the rainbow of ugly stamp marks you were meant to be! Or deny the greytide and mark them for it! You can even approve of beneficial members of your team as head, giving them a sign of your appreciation. Its as harmless as a walk to the showers and brings some more charm to the game.

## Images of changes
KCr2xM2hvr

## Changelog
:cl:
add: Adds the ability to stamp people.
imageadd: Adds marks that display on stamped people.
/:cl:

### CODE REVIEW

**Comment by Fox-McCloud** [comment]
Quite frankly, adding an entire layer to render, just for a gimmick like this, isn't warranted at all.

**Comment by kugamo** [comment]
> 
> 
> Quite frankly, adding an entire layer to render, just for a gimmick like this, isn't warranted at all.

Like I said in the original PR, I have since changed the code into using the existing MISC layer. This PR does not add a new render layer for humans.

**Comment by PacifistDalek** [comment]
Heck yes

**Comment by Oyu-De** [comment]
> Heck yes. Time to give clowns a rainbow stamp

Clown: Look security, I AM stamped and have all access. "honk"

**Comment by kenken244** [comment]
think this counts as a bug fix, so... speedmerge pls?

File: code/modules/mob/living/carbon/human/examine.dm
Comment (AffectedArc07) [inline code/modules/mob/living/carbon/human/examine.dm:175]: See about using `english_list()` here. It does this for you and looks cleaner.

File: code/modules/paperwork/stamps.dm
Comment (AffectedArc07) [inline code/modules/paperwork/stamps.dm:27]:
```suggestion
	if(length(H.stamp_marks.overlays) >= 50)	//no more than 50 marks on a person, otherwise infinite laggy stamp marks
```
Dont want someone somehow getting 51 then applying infinite amounts. Also 50 still seems way too high (Keep in mind a lot of the people playing this game dont have hyper-powerful computers, and 50 overlays is technically processing an extra 51,200 pixels.

File: code/modules/paperwork/stamps.dm
Comment (AffectedArc07) [inline code/modules/paperwork/stamps.dm:33]: How does this react to when you click the icon in the alt+click panel in the statpane? Those screw with `icon-` params and it could result in stamp marks way off the sprite.

**Review by AffectedArc07 — CHANGES_REQUESTED** [review]

**Comment by variableundefined** [comment]
Not keen on this PR after what happened yesterday. 

Objection.

**Comment by Pixelzilla** [comment]
I just don't understand how this is something we really need. It just seems like it is going to be used to antagonize people easily and quickly with no real way to avoid being stamped by someone who is intent on annoying people. Sure, you can go and shower to remove it but some of us want to actually play the game, rp and not have to deal with what i am sure will be a feature abused by people who's primary source of fun is at the expense of others.

Dont we already have issues with sprite layers fully rendering sometimes as it is? I cant imagine adding a ton of stamp overlays is going to help and if we're going to add more layers to mobs i think it would be much better suited to something that is a player's customization choice and not a tool that is going to just cause frustration for most people who it is forced upon.

Considering how frustrating it is to have the barber to change your hair/hair color/fur color, heck even genetics force injecting obnoxious UI. All of those things are very irritating to have happen against your will but at least they take time and cant rapidly be done yet the also have positive uses...i just don't really see anything positive in this pr - it really seems like it will primarily be used maliciously on people who probably wont appreciate it.

File: code/modules/paperwork/stamps.dm
Comment (kugamo) [reply code/modules/paperwork/stamps.dm:33]: dreamseeker_peeKgjs1tY
I believe because they are icons that are initially offset then overlayed on an image and then finally that is overlayed on a mutable appearance in update_icons.dm it doesn't respond poorly? Only issue I see with them in that menu is that the coloration of them is lost, which although an issue Its one thats shared with all bloody sprites for humans.

File: code/modules/paperwork/stamps.dm
Comment (kugamo) [reply code/modules/paperwork/stamps.dm:27]: I dropped it down to 20, which ought to cut that number in half at least. I have made the suggested changes to insure that if somehow more are added the code addresses that.

**Comment by kugamo** [comment]
> 
> 
> Not keen on this PR after what happened yesterday.
> 
> Objection.

Although I cannot promise nothing will go wrong with it being merged, I can confidently say I have taken measures to prevent any. I dont know what went wrong exactly, nor am I aware of anyone possessing that knowledge. I suspect it was some of my commits on the last PR, not all of them went Travis approved but I though removing the code with a new commit would remove the issue. _Apparently not._ To address this issue I condensed my code into a single commit. I also had discrepancies on my fork at the time, which that PR was based on. I have since made my fork even with the master, and newly created this code based on that updated version. Looking at both of these efforts I hope you reconsider. I accept that if this PR fails scuff it would be proper to deny it, but without knowing what exactly went wrong I would be regretful if that happened. 

> 
> 
> I just don't understand how this is something we really need. It just seems like it is going to be used to antagonize people easily and quickly with no real way to avoid being stamped by someone who is intent on annoying people. Sure, you can go and shower to remove it but some of us want to actually play the game, rp and not have to deal with what i am sure will be a feature abused by people who's primary source of fun is at the expense of others.

If this is as big an issue as you paint it to be here there are things that could be added to it to make it 'resistable', maybe a action timer so it takes a second to do. But I honestly think it will not be too terrible, the icons are like 8 pixels, and stamps are pretty uncommon outside of cargo. Apart from traitors, cargo, the clown and the heads noone has access to them by default. Apart from this they are very easily washed off, soap or shower, which are far from uncommon. 

>
> 
> Dont we already have issues with sprite layers fully rendering sometimes as it is? I cant imagine adding a ton of stamp overlays is going to help and if we're going to add more layers to mobs i think it would be much better suited to something that is a player's customization choice and not a tool that is going to just cause frustration for most people who it is forced upon. 
>i just don't really see anything positive in this pr - it really seems like it will primarily be used maliciously on people who probably wont appreciate it.

To be clear this PR doesnt add more layers to humans, It does add a new image to humans and up to 20 stamp icons. Although possibly alot I doubt it will cause any severe issues rendering wise. _(and I am perfectly willing to try condensing that number if it proves otherwise)_ I think you are looking at this from only the side of the stamped, and not from the side of the stamper. Although inconvenient for the stamped its easy as tapping yourself with a soap or walking to the showers to resolve it, and its 8 pixels per mark that only show on one side of your character. I am sure it could even be worn as a badge of honor for the devoted stamp collecting player. And for the stamper its a way to mark people for their actions, or maybe as a prank which can be fun. Altogether I think it adds more interactivity to the game and to objects that go underused outside of the bureaucracy nuts, and in a fun and harmless way.

File: code/modules/paperwork/stamps.dm
Comment (AffectedArc07) [reply code/modules/paperwork/stamps.dm:33]: I mean clicking the mob in that image with the stamp in hand, the `icon-x` and `icon-y` values get really screwed up.

Comment (kugamo) [reply code/modules/paperwork/stamps.dm:33]: At worst they wont show up on the character and they will have their descriptions updated, my code cuts out any marks that won't appear on a base person with their clothes, hair, and tail. But from my testing it works very well.

File: code/modules/paperwork/stamps.dm
Comment (SteelSlayer) [inline code/modules/paperwork/stamps.dm:28]: 1) Comment was mismatched from the code
2) IMO a reason other than "Bluespace mechanics" would be nice here. But not a huge deal either way.
```suggestion
	if(length(H.stamp_marks.overlays) >= 20)	//no more than 20 marks on a person, otherwise infinite laggy stamp marks
		to_chat(user, "<span class='notice'>[target.name] is already completely covered in stamps!</span>")
```

Comment (kugamo) [reply code/modules/paperwork/stamps.dm:28]: I changed it to "[target.name]'s skin is saturated with ink! No more marks can appear on them." , hopefully this is less 4th wall breaking.

**Comment by Spacemanspark** [comment]
I don't know how I feel about this.

On one hand, this is totally something I'd run around doing as Spark.

On the other hand... I can imagine this is just going to be annoying for players to deal with.

File: code/modules/paperwork/stamps.dm
Comment (AffectedArc07) [inline code/modules/paperwork/stamps.dm:28]: Why the extra `'` after `them.`

Comment (kugamo) [reply code/modules/paperwork/stamps.dm:28]: dumb mistake

**Comment by Pixelzilla** [comment]
> If this is as big an issue as you paint it to be here there are things that could be added to it to make it 'resistable', maybe a action timer so it takes a second to do. But I honestly think it will not be too terrible, the icons are like 8 pixels, and stamps are pretty uncommon outside of cargo. Apart from traitors, cargo, the clown and the heads noone has access to them by default. Besides this they are very easily washed off, soap or shower, which are far from uncommon.

In a game where many players seem to make it their primary goal to annoy other players as much as physically possible and where we have a role (Clown) dedicated to basically being irritating you're saying you don't see how it can be used maliciously to disrupt other players who are just trying to RP and enjoy the game?

The icons may be "small" but from the example gif you've shown that you can stamp people multiple times in many locations and effectively make them look like a rainbow vomit monster. Stamps are far from uncommon. They are easily obtained in cargo, maint loot, the clown, IAA, Magistrate, all heads have a stamp too. There are more stamps at round start than there are bars of soap.

> To be clear this PR doesnt add more layers to humans, It does add a new image to humans and up to 20 stamp icons. Although possibly alot I doubt it will cause any severe issues rendering wise. _(and I am perfectly willing to try condensing that number if it proves otherwise)_ I think you are looking at this from only the side of the stamped, and not from the side of the stamper. Although inconvenient for the stamped its easy as tapping yourself with a soap or walking to the showers to resolve it, and its 8 pixels per mark that only show on one side of your character. I am sure it could even be worn as a badge of honor for the devoted stamp collecting player. And for the stamper its a way to mark people for their actions, or maybe as a prank which can be fun. Altogether I think it adds more interactivity to the game and to objects that go underused outside of the bureaucracy nuts, and in a fun and harmless way.

I'm not a coder so when i'm referring to it adding additional "layers" i mean you're adding more images to a mob sprite and apparently there are a lot of potential images. Even if they are "only" 8 pixels player sprites are only 32 x 32. Those 8 pixels add up.

While you're right that I'm not approaching this from the perspective of the person stamping that's because personally I would probably feel bad using this on other people because I know I personally would not like it being used on me. I would not want to stop whatever interaction I was having to have to go locate soap or a shower because some random person drive-by stamped me (probably multiple times because why wouldn't you do it until you cant anymore?). In the past I did enjoy using stamps on people when all it did was a message but if it actually left a mark i probably wouldn't do it at all. I do believe that most of the appeal of this from the "stamper's perspective" will be at the expense of other players. I doubt the majority will find it funny or positive when it happens to them.

**Comment by kugamo** [comment]
>
> 
> In a game where many players seem to make it their primary goal to annoy other players as much as physically possible and where we have a role (Clown) dedicated to basically being irritating you're saying you don't see how it can be used maliciously to disrupt other players who are just trying to RP and enjoy the game?

I can see that some players might use it to annoy people, but honestly theres rules against this on our server. Its not like that behavior goes unpunished or isnt looked down upon in our community, I maybe speaking for myself here but I dont think this PR  explicitly encourages it. Besides, there are active measures in our community against it. There are already plenty of other ways to disrupt well meaning RP and I havent seen it as a large as a problem as you have made it out to be here.

>
>
> While you're right that I'm not approaching this from the perspective of the person stamping that's because personally I would probably feel bad using this on other people because I know I personally would not like it being used on me. I would not want to stop whatever interaction I was having to have to go locate soap or a shower because some random person drive-by stamped me (probably multiple times because why wouldn't you do it until you cant anymore?). In the past I did enjoy using stamps on people when all it did was a message but if it actually left a mark i probably wouldn't do it at all. I do believe that most of the appeal of this from the "stamper's perspective" will be at the expense of other players. I doubt the majority will find it funny or positive when it happens to them.

Believe it or not SS13 is not just a RP game where everyone is a civilian and sits around and talks, we have antags as well as jobs that are made so you play a role on the station working all in their own ways. As un-ideal as it is to go and take a shower or find some soap its not like this cannot be RPed around as well, nor does it make you a useless victim against malicious trouble makers. Maybe in the scenario I first brought up stamping marks on people wouldnt be so great, but that isnt the scenario on the server as it currently exists. With antagonists and different roles we are all RPing in a pre-structured environment, with bad actors, good ones and idiotic ones all with their own motivations. If you are so against this PR because it takes you out of the Ideal RP scenario I have to think why you dont advocate for the removal of all antags. It can be plenty fun to have alitte dig on someone for whatever reason, and if its a mostly harmless one all the better. How many times have you shoved or disarmed someone for being annoying or being a shitter, this is all but akin to that and in a arguably less harmful way. 

I don't expect you to wildly change your mind on all this, but I think its worth the effort to do due justice to the virtues of my PR, Which I still believe it has. Honestly I dont expect it to be used only to antagonize well meaning RPers, but I can see how it could be. But if you ask me, what you have argued so far does too little justice to common players's intentions, and to the ruled community we play in. If there are malicious actors out there that would use this mechanic to abuse other players I expect them to be handled like all other ones, with a bwoink in regard to rule 0.

I only wish to add to what fun we can have together on this server, and not a one sided sadistic fun let me be clear. I can think of plenty of ways this mechanic of stamping people can be used in a wholesome ways, or at least in ways inline with the structured play environment as it currently exists on the server today. Be it mostly harmless pranks, or retribution for bad actors,  a somewhat abusive approval, or maybe even as a minigame for the avid stamp collecting clown. Its a tool at the end of the day that I think adds to the ways in which we can have fun together on the server, and if its alittle troublesome so be it...

**Comment by kugamo** [comment]
I have lost interest in maintaining this PR, and if I am being honest, unhappy with its neglect. I am tired of coming back to this page to see nothing has changed. I privately messaged multiple times the maintainer who objected for clarification and was ignored, and the other persists in their objection despite the reasons for it being outside of my control and my attempt to attend to them. I realize a contested PR takes more time to process than a fully approved one and it seems this is going continue to be one despite my best efforts. 1/2 a month later I just want it off my mind more than I care about seeing it in game.

**Comment by Spacemanspark** [comment]
If people don't like something, they have a right to not like it. I don't think it's fair to blame them for giving critism or their thoughts on it.
As for the maintainers, they have a heavier workload on their shoulders than most give them credit for. This is a relatively minor PR with little interest from many. It makes sense that more important matters or bigger features will take precedent over this. Not to mention that maintainers are volunteers. They haven't treated you poorly, and they have lives as well. I find you to be acting unfair towards them, and I hope you consider all angles when you conduct future PRs. We're all only human.

**Comment by kugamo** [comment]
@Spacemanspark 
> 
> 
> If people don't like something, they have a right to not like it. I don't think it's fair to blame them for giving criticism or their thoughts on it.

When did I ever blame anyone for not liking this PR? I am perfectly fine with people not liking my PRs and with them speaking their minds. Most of the time its constructive and contributes to better code.

>
>
> As for the maintainers, they have a heavier workload on their shoulders than most give them credit for. This is a relatively minor PR with little interest from many. It makes sense that more important matters or bigger features will take precedent over this. Not to mention that maintainers are volunteers. They haven't treated you poorly, and they have lives as well. I find you to be acting unfair towards them, and I hope you consider all angles when you conduct future PRs. We're all only human.

I never asked any maintainers to review my PR or change their opinion of it. I only reached out for some clarification of what they meant, and was ignored by one of them. Yeah I understand they can be busy, but it was 3 months of not one message back. The other after a codebase error with merge attempt 1 didnt talk to me once about what went wrong or how I could fix it, and then after I took steps on my own to fix it they objected. If thats not being treated poorly I don't know what is.

If you think me closing a PR is acting unfair then I don't know what to say to you, I have been nothing but cordial up to now. As to you belittling my PR and making me out to have bad intentions where I have none, I think you ought to apologize.

**Pull request closed without merging**
